classlib/src/main/java/org/teavm/classlib/java/util/TMap.java

```diff
@@ -294,4 +294,8 @@ static <K, V> TMap<K, V> ofEntries(TMap.Entry<K, V>... entries) {
     static <K, V> TMap.Entry<K, V> entry(K k, V v) {
         return new TTemplateCollections.ImmutableEntry<>(requireNonNull(k), requireNonNull(v));
     }
+
+    static <K, V> TMap<K, V> copyOf(TMap<? extends K, ? extends V> map) {
```

Comment:
I believe `NEntriesMap` can override this map to simply return itself without copying.

Reply:
I don't think I fully understand. The API of `copyOf()` specifies: 

> Returns an unmodifiable Map containing the entries of the given Map. The given Map must not be null, and it must not contain any null keys or values. If the given Map is subsequently modified, the returned Map will not reflect such modifications.

I interpreted that as: If for example a `HashMap` is passed the `copyOf` method must return a new immutable map. 

I could add a check to see if the passed argument is already an (unmodifiable) `NEntriesMap` instance. In which case we can simply return that. Is that what you meant?

Reply:
> Is that what you meant?

Yes, that's what I mean. However, you don't need to check anything with `instanceof`, let polymorphism work for you. Just write specific implementation in an immutable implementation.

Reply:
Thanks for the clarification! However I don't think using polymorphism will work for this particular case. As far as I know (granted, that's a low bar ;-) this can in practice only be implemented with `instanceof`. For example adding this method to `Maps`:

```java
static <K, V> TMap<K, V> copyOf(TTemplateCollections.NEtriesMap<K, V> map) {
    return map;
}
```

and then using a test such as below will fail:

```java
@Test
public void copyOfOptimized() {
    Map<String, Integer> map = Map.of("q", 0, "w", 1, "e", 2);
    Map<String, Integer> mapCopy1 = Map.copyOf(map);
    Map<String, Integer> mapCopy2 = Map.copyOf(mapCopy1);

    assertSame("Must not create copies of immutable collections", mapCopy1, mapCopy2);
}
```

Because the method overloading will only select the `copyOf(TTemplateCollections.NEtriesMap<K, V> map)` method if the caller explicitly calls the method with a `NEtriesMap` type. However in practice this won't happen since the caller should never be aware they are using the `NEtriesMap` implementation of the `Map` interface. 

Changing the `copyOf` implementation to something like this does pass the test:

```java
@SuppressWarnings("unchecked")
static <K, V> TMap<K, V> copyOf(TMap<? extends K, ? extends V> map) {
    if (map instanceof TTemplateCollections.NEtriesMap) {
        return (TTemplateCollections.NEtriesMap<K, V>) map;
    } else {
        return new TTemplateCollections.NEtriesMap<>(map);
    }
}
```

Reply:
Ah, did not notice `static` modifier. Then you are right, there's no way to apply polymorphism here.

Reply:
I've updated the PR to prevent it from copying `NEtriesMap` instances. I've also added a test for this to ensure this behavior doesn't silently break in the future.